pkg/provider/cloud/openstack/provider.go

```diff
@@ -195,233 +199,398 @@ func validateExistingSubnetOverlap(networkID string, netClient *gophercloud.Serv
+			return nil, err
 		}
 	}
-	if cluster.Spec.Cloud.Openstack.IPv6SubnetID != "" && cluster.Spec.Cloud.Openstack.RouterID == "" && routerID == "" {
-		var err error
-		routerID, err = getRouterIDForSubnet(netClient, cluster.Spec.Cloud.Openstack.IPv6SubnetID)
+	// All machines will live in one dedicated security group.
+	if force || cluster.Spec.Cloud.Openstack.SubnetID == "" || cluster.Spec.Cloud.Openstack.IPv6SubnetID == "" {
+		cluster, err = reconcileSubnet(ctx, netClient, cluster, update, os.dc.DNSServers)
 		if err != nil {
-			return nil, fmt.Errorf("failed to verify that the subnet '%s' has a router attached: %w", cluster.Spec.Cloud.Openstack.IPv6SubnetID, err)
+			return nil, err
 		}
 	}
-	// If router has to be created, add associated finalizer.
-	if cluster.Spec.Cloud.Openstack.RouterID == "" && routerID == "" {
-		finalizers = append(finalizers, RouterCleanupFinalizer)
+	if force || cluster.Spec.Cloud.Openstack.RouterID == "" {
+		cluster, err = reconcileRouter(ctx, netClient, cluster, update)
+		if err != nil {
+			return nil, err
+		}
 	}
+	return cluster, nil
+}
 
-	if ipv4Network {
-		// If subnet has to be created, add associated finalizer.
-		if cluster.Spec.Cloud.Openstack.SubnetID == "" {
-			finalizers = append(finalizers, SubnetCleanupFinalizer)
+func reconcileNetwork(ctx context.Context, netClient *gophercloud.ServiceClient, cluster *kubermaticv1.Cluster, update provider.ClusterUpdater) (*kubermaticv1.Cluster, error) {
+	networkName := cluster.Spec.Cloud.Openstack.Network
+
+	// If the network name is set, check if it exists
+	if networkName != "" {
+		existingNetwork, err := getNetworkByName(netClient, networkName, false)
+		if err != nil {
+			if !isNotFoundErr(err) {
+				// If there's an error other than the network not being found, return it
+				return nil, fmt.Errorf("failed to check for existing network: %w", err)
+			}
 		}
-		// If subnet or router has to be created, subnet needs to be attached to the router.
-		if cluster.Spec.Cloud.Openstack.SubnetID == "" || (cluster.Spec.Cloud.Openstack.RouterID == "" && routerID == "") {
-			finalizers = append(finalizers, RouterSubnetLinkCleanupFinalizer)
+		if existingNetwork != nil {
+			return cluster, nil
 		}
 	}
-	if ipv6Network {
-		// If subnet has to be created, add associated finalizer.
-		if cluster.Spec.Cloud.Openstack.IPv6SubnetID == "" {
-			finalizers = append(finalizers, IPv6SubnetCleanupFinalizer)
+	network, err := getNetworkByName(netClient, resourceNamePrefix+cluster.Name, false)
+	if err != nil {
+		if !isNotFoundErr(err) {
+			return nil, fmt.Errorf("failed to check network: %w", err)
 		}
-		// If subnet or router has to be created, subnet needs to be attached to the router.
-		if cluster.Spec.Cloud.Openstack.IPv6SubnetID == "" || (cluster.Spec.Cloud.Openstack.RouterID == "" && routerID == "") {
-			finalizers = append(finalizers, RouterIPv6SubnetLinkCleanupFinalizer)
+	} else {
+		if network != nil {
```

Comment:
This if statement is redundant, you control the getNetworkByName and if that function returns no error _and_ no network, the code should explode (panic) to indicate an issue with the getNetworkByName function. Adding an if here is just hinding potential bugs.